File: 프로그래머스/2/42746. 가장 큰 수/가장 큰 수.java
/*
- String 도 정렬 가능
- 더 큰 수 Arrays.sort(arr, (o1, o2) -> (o2 + o1).compareTo(o1 + o2));
- 0 으로 시작할 경우 고려!!
*/
import java.util.*;
class Solution {
    public String solution(int[] numbers) {
        String[] arr = new String[numbers.length];
        for (int i = 0; i < arr.length; i++) {
            arr[i] = String.valueOf(numbers[i]);    
        }
        
        Arrays.sort(arr, (o1, o2) -> (o2 + o1).compareTo(o1 + o2));
        
        String answer = "";
        
        for (int i = 0; i < arr.length; i++) {
            answer += arr[i];
        }
        
        if (answer.startsWith("0")) {
            return "0";
        }
        
        return answer;
    }
}

File: 프로그래머스/2/42746. 가장 큰 수/가장 큰 수.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

bool compare(string a, string b) {
    return a + b > b + a;
}

string solution(vector<int> numbers) {
    vector<string> strvec;
    for (int n : numbers) strvec.push_back(to_string(n));
    
    sort(strvec.begin(), strvec.end(), compare);
    
    // 가장 큰 숫자부터 정렬했는데도 첫 번째가 "0"이면 → 전부 0이라는 뜻이므로, "0"만 리턴
    if (strvec[0] == "0") return "0";
    
    string answer = "";
    for (string s : strvec) answer += s;
    return answer;
}